video/out/opengl/context_drm_egl.c

```diff
@@ -498,16 +498,27 @@ static void queue_flip(struct ra_ctx *ctx, struct gbm_frame *frame)
     data->waiting_for_flip = &p->waiting_for_flip;
     data->log = ctx->log;
 
+    int flags = DRM_MODE_PAGE_FLIP_EVENT | DRM_MODE_ATOMIC_NONBLOCK;
+    if(ctx->vo->opts->drm_opts->drm_hdr_metadata) {
+        ret = drm_create_hdr_metadata(atomic_ctx, &p->hdr_data);
+        if (ret) {
+            MP_WARN(ctx->vo, "Failed to create HDR metadata (%d)\n", ret);
+        } else if (p->hdr_data.blob_id != 0) {
+            drm_object_set_property(atomic_ctx->request, atomic_ctx->connector,
+                                    "HDR_OUTPUT_METADATA", p->hdr_data.blob_id);
+            flags |= DRM_MODE_ATOMIC_ALLOW_MODESET;
```

Comment:
Why do we create/attach the metadata in the `queue_flip` function, instead of during the initial setup in `crtc_setup_atomic`? That doesn't feel right.

Internally the kernel disables the `allow_modeset` during pageflips, so I'm not 100% sure if/how this will work.

Reply:
My bad - didn't know the mpv VO architecture. Having a quick look now, it seems that:
 - init/uninit are called on mpv startup and close, respectively
 - reconfig is called as a source is opened and ultimately the window is resized

Looking the reconfig implementations - wayland, x11 and drm:
 - wayland/x11 have a window
 - drm does not - the closest equivalent is a hardware plane
 - wayland/x11 the window is created/resized, during the reconfig hook
 - drm the "window" is setup during "init"
 - wayland/x11 handle resizes (be that manual or full-screen)
 - drm always uses full-screen, pressing F does nothing

So unless I'm missing something, the drm code should be reshuffled a bit to behave like the others. Will give that quick go tomorrow.

With that fixed, we should be able to move the HDR handling into the reconfig hook - instead of doing it on each flip.

Reply:
If you're looking to do some refactoring, one of the first things I was going to to was to de-duplicate a lot of code in `context_drm_egl.c` and `vo_drm.c`, it's doing a lot of the same things slightly differently and that seems ripe for refactoring. This could potentially be a part of that as well. I was going to take a stab at that eventually as well. If you stop by in `#mpv-devel` we can go over the specifics.